### ISSUE TITLE
Training loop from scratch for multi-out model - how to calculate loss and update it.

### ISSUE THREAD

**Issue by Suzan009** [tensorflow/tensorflow#43580]
I have a ML models which have 3 output. Each of them have passed into a Soft-Max player. I'm trying to train the model from scratch with training loop in tf.keras. My generator output are something like this:

```
x,y = (batch_size, h,w,c), [(batch_size, 2), (batch_size, 8), (batch_size, 10)]
```

As you notice, each of the 3 output has different numbers of labels. First output layer has 2 labels output, second one has 8 and last one has 10. The problem is not about the data generator; It's fine. **However, I don't understand how will I calculate loss function and update it in scratch training.**

---

I've started with [document guidelines](https://keras.io/guides/writing_a_training_loop_from_scratch/). Here is the code snippet,

```python

epochs = 1
optimizer   = tf.keras.optimizers.Adam(learning_rate=0.05)
loss_fn     = tf.keras.losses.CategoricalCrossentropy()

train_acc_metric = tf.keras.metrics.Accuracy()
val_acc_metric   = tf.keras.metrics.Accuracy()

@tf.function
def train_step(x, y):
    with tf.GradientTape() as tape:
        logits = model(x, training=True)  # Logits for this minibatch

        # Compute the loss value for this minibatch.
        train_loss_value = loss_fn(y, logits)

    grads = tape.gradient(train_loss_value, model.trainable_weights)
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
    train_acc_metric.update_state(y, logits)
    return train_loss_value
```

And, 

```python

for epoch in range(epochs):
    # Iterate over the batches of the dataset.
    for step, (x_batch_train, y_batch_train) in enumerate(train_generator):
        train_loss_value = train_step(x_batch_train, y_batch_train)

    # Reset metrics at the end of each epoch
    train_acc_metric.reset_states()
```

I suspect, I am not handling properly the computation of loss in here: `loss_fn(y, logits)`. `y` has 3 output, `logits` has also 3 output. 

Using `model.compile` is simply straight-forward, just passing the loss function in the loss argument. But what to do in scratch training. The 3 output (though are co-related) needs to trained separately. 

Asked [SO](https://stackoverflow.com/q/64065473/9215780), nobody knows.

**Comment by Suzan009** [comment]
The `logits = model(x, training=True)` has 3 output. So to calculate the loss function separately (I think) we need to unpack it. `loss_one, loss_two, loss_thr = model(x)`. And next, maybe like this:

```
train_loss_value_one = loss_fn(y[0], loss_one)
train_loss_value_two = loss_fn(y[1], loss_two)
train_loss_value_thr = loss_fn(y[2], loss_thr)
```
I'm feeling I'm definitively missing something.

**Comment by Suzan009** [comment]
@ravikyram is it possible in custom training in tf.keras?

**Comment by jvishnuvardhan** [comment]
@Suzan009 Yes, is it possible in custom training in `tf.keras`. Are you facing any error? Can you please share a standalone code to reproduce the error. Thanks!

**Comment by Suzan009** [comment]
Mainly I need to solve above scenario. I know we can do custom training in tf.keras. But how to define loss and metrics for multi-label multi class classification problem.

Please, consider the following cases:

```
x,y = (batch_size, h,w,c), [(batch_size, 2), (batch_size, 8), (batch_size, 10)]
```

Here, the model has 3 output. But each of the single output layer doesn't give single out put itself. See, first output has 2 different classes, second one has 8 different classes and third one has 10 different classes. Now each of these three output are mutually related. 

In such problem, how will I define the loss and metrics in custom training in tf.keras. 

I can do it in old keras. fairly simple. Such as:

```python
oputput1 = Dense(2,  activation='softmax') (curr_output)
oputput2 = Dense(8,  activation='softmax') (curr_output)
oputput3 = Dense(10,  activation='softmax') (curr_output)
output_tensor = [oputput1, oputput2, oputput3]

model = Model(input_tensor, output_tensor)
```
And compile like this, 

```python

model.compile(
optimizer = , 
loss = 'categorical_crossentropy',
metrics='accuracy'
)
```

**Comment by chhavijain2212** [comment]
Hey,
Answering wrt TF2.4
You will need a separate metric for each output task. While you can calculate the total epoch loss by adding the three losses, the accuracy for each will need to be calculated separately only.

`
train_acc_metric_label1 = tf.keras.metrics.Accuracy()
train_acc_metric_label2 = tf.keras.metrics.Accuracy()
train_acc_metric_label3 = tf.keras.metrics.Accuracy()
`

You have already reached this point.
> 
> 
> The `logits = model(x, training=True)` has 3 output. So to calculate the loss function separately (I think) we need to unpack it. `loss_one, loss_two, loss_thr = model(x)`. And next, maybe like this:
> 
> ```
> train_loss_value_one = loss_fn(y[0], loss_one)
> train_loss_value_two = loss_fn(y[1], loss_two)
> train_loss_value_thr = loss_fn(y[2], loss_thr)
> ```
> 
> I'm feeling I'm definitively missing something.

Now, you can simply update them as following
`
logits = model(x, training=True)
train_acc_metric_label1.update_state(y[0], logits[0])
train_acc_metric_label2.update_state(y[1], logits[2])
train_acc_metric_label3.update_state(y[1], logits[2])
`
For total loss value that you are trying to return from `train_step()`
`
train_loss_total = loss_fn(y[0], logist[0]) + loss_fn(y[1], logist[1]) + loss_fn(y[2], logist[2])
`
